# Problema
# 1. En el libro Numerical Mathematics and Computing de Cheney y Kincaid, observe el ejercicio 4, literal a de la página 288 (el libro se encuentra en el enlace proporcionado en la sección Sitios).
# 2. Utilice el siguiente código para resolver el ejercicio (utilizando un sistema de cien ecuaciones y cien incógnitas). El código es del mismo libro.

In [1]:
import numpy as np
def tridiag(n,d,a,c,b):
    x = np.zeros(n)
    for i in range(1,n-1):
        mult = a[i-1]/d[i-1]
        d[i] -= mult*c[i-1]
        b[i] -= mult*b[i-1]
    print(d)
    x[n-1] = b[n-1]/d[n-1]
    for i in range(n-2,-1,-1):
        x[i] = (b[i]-c[i]*x[i+1])*d[i]
    return x

# Use el procedimiento Tri para resolver el siguiente sistema de 100 ecuaciones. Compare la solución numérica con la obvia solución exacta.
\\begin{cases}
\hphantom{-}x_1 + 0.5x{_2} \qquad\qquad = 1.5 & \\
0.5x_i{_-{_1}} + x_i + x_i{_+{_1}} \quad= 2\qquad(2\leq j\leq n-1) & \\
\qquad 0.5x_9{_9}+x_1{_0{_0}} \quad= 1.5
\end{cases}

In [2]:
n=100 # número de ecuaciones
a=0.5*np.ones([100]) # vector subdiagonal con todos sus valores en 0.5
b=2*np.ones([100]) # vector de términos independientes con todos sus valores en 2
b[0]=b[99]=1.5 # corrijo los extremos del vector de términos independientes
c=a # vector superdiagonal igual al vector subdiagonal
d=np.ones([100]) # vector diagonal con todos sus valores en 1
print("Vector superdiagonal\n\n",c,"\n")
print("Vector diagonal\n\n",d,"\n")
print("Vector subdiagonal\n\n",a,"\n")
print("Vector de términos independientes\n\n",b,"\n")
tridiag(n,d,a,c,b)

Vector superdiagonal

 [0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5
 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5
 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5
 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5
 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5
 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5] 

Vector diagonal

 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1.] 

Vector subdiagonal

 [0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5
 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5
 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5 0.5
 0

array([1.14314355, 0.7137129 , 0.59676561, 0.54303651, 0.51228316,
       0.49238946, 0.47847422, 0.46819738, 0.46029819, 0.45403764,
       0.44895405, 0.44474423, 0.44120081, 0.43817722, 0.43556694,
       0.43329065, 0.43128813, 0.4295128 , 0.42792809, 0.42650485,
       0.42521962, 0.42405324, 0.42298997, 0.42201671, 0.4211225 ,
       0.42029808, 0.41953559, 0.41882831, 0.41817045, 0.41755701,
       0.41698363, 0.41644651, 0.41594232, 0.41546813, 0.41502133,
       0.41459961, 0.41420092, 0.41382343, 0.41346549, 0.41312562,
       0.41280248, 0.41249486, 0.41220169, 0.41192195, 0.41165475,
       0.41139927, 0.41115475, 0.4109205 , 0.41069588, 0.41048032,
       0.41027327, 0.41007425, 0.40988279, 0.40969847, 0.4095209 ,
       0.40934972, 0.40918458, 0.40902517, 0.40887121, 0.40872241,
       0.40857851, 0.40843929, 0.40830452, 0.40817398, 0.40804748,
       0.40792484, 0.40780588, 0.40769044, 0.40757836, 0.4074695 ,
       0.40736373, 0.40726091, 0.40716092, 0.40706365, 0.40696

# Comprobación

$$x_1 + 0.5x{_2} \qquad\qquad = 1.5$$

$$Solución\quad numérica$$

$$1 + 0.5*0.75$$

$$1 + 0.375 = 1.375 \neq 1.5$$

$$Solución\quad exacta$$

$$1.14314355 + 0.5*0.7137129$$

$$1.14314355 + 0.35685645 = 1.5 = 1.5$$